Add multiple filters, custom conditions

lf- opened the pull request:
This is a breaking change for existing installations, and I bumped the version accordingly.

Exact syntax absolutely up to bikeshedding, however, the tilde separator was chosen as it would be very uncommon in event names, regexes, *and* can be included in URLs without any URL encoding.

I initially needed this change set to be able to invert conditions, but figured since I needed to hack on it anyway, I could make the matching powerful.

lf- committed:
Add multiple filters, custom conditions

lf- commented:
Also, there's a bug I found in the existing code (not related to my PR): it eats event descriptions (annoying but tolerable) and interprets all-day events as 0 minute long events at the day, GMT. I will investigate and probably fix this and/or event descriptions soon and am just writing this down so I remember.

Wassasin commented:
Very cool! The MR looks great!

Would you consider updating the error handling to be more inline with the rest of the crate? I've shied away from using `Box<dyn std::error::Error>` when it is not necessary.

Concerning the API: perhaps the tilde separator could be replaced with a more `application/x-www-form-urlencoded` native solution, like a list of filters. I.e. `?url=http://test.com/ical&filter[]=equals:first&filter[]=!contains:second`. Also we could consider reverting to equals (or contains?) when no operator is included, only text (i.e. no `:`). Requiring an explicit operator also has its merits though.

lf- commented:
I considered the approach of using filter[] but couldn't figure out how to do it after trying for a while. If you know more about serde and the web crates, I'd appreciate some help on this.

As for accepting filters with no operator, I judged it more surprising to have a filter be accidentally reinterpreted as soon as a colon is added (as is often in event summaries) than rejecting it. Maybe I could resurrect the filter URL parameter (singular) so it adds a contains filter with just the content of the parameter?

lf- committed:
Refactor error handling, add socket activation

Socket activation is necessary to make it easier to use systemd to
sandbox this service, since it's basically one big SSRF vector if not
sandboxed away from localhost and any local services whatsoever.

Also adds a hack per
https://github.com/rust-lang/cargo/issues/8075#issuecomment-610259732
so you can run `cargo clippy --features clippy`.

lf- commented:
@Wassasin I've refactored the error handling as requested and added socket activation so I can sandbox the service completely away from localhost access (since it's a big SSRF vector otherwise).

If you have any ideas for how to implement the changes to the URL format, I'm still not sure how to do that.

Wassasin committed:
Use urlencoded sequences for filter instead of tilde

Wassasin committed:
Removed superfluous 'true' filter

Wassasin commented:
@lf- Sorry that it took a while before I could get to this. Could you check the commits I made to see if you agree? If so I'll merge this MR asap.

lf- commented:
Ah yeah, you're right, the True filter was superfluous since .all() is true for an empty iterator. Nice. Assuming it's tested and confirmed working, this looks good to go.
